op-challenger/challenger/traversal.go

```diff
@@ -0,0 +1,238 @@
+		err := backoff.DoCtx(ctx, 5, backoffStrategy, func() error {
+			ctx, cancel := context.WithTimeout(ctx, time.Second*10)
+			defer cancel()
+			receipt, err := l.client.TransactionReceipt(ctx, tx.Hash())
+			receipts = append(receipts, receipt)
+			return err
+		})
+		if err != nil {
+			return nil, err
+		}
+	}
+	return receipts, nil
+}
+
+// subscribeNewHead subscribes to new heads with a backoff.
+func (l *logTraversal) subscribeNewHead(ctx context.Context, headers chan *types.Header) (ethereum.Subscription, error) {
+	backoffStrategy := l.buildBackoffStrategy()
+	var sub ethereum.Subscription
+	err := backoff.DoCtx(ctx, 4, backoffStrategy, func() error {
+		ctx, cancel := context.WithTimeout(ctx, time.Second*5)
+		defer cancel()
+		var err error
+		sub, err = l.client.SubscribeNewHead(ctx, headers)
+		return err
+	})
+	return sub, err
+}
+
+// Start starts the log traversal.
+func (l *logTraversal) Start(ctx context.Context, done chan struct{}, handleLog func(*types.Log) error) error {
+	if l.started {
+		return ErrAlreadyStarted
+	}
+	headers := make(chan *types.Header)
+	sub, err := l.subscribeNewHead(ctx, headers)
+	if err != nil {
+		l.log.Error("failed to subscribe to new heads", "err", err)
+		return err
+	}
+	go l.onNewHead(ctx, headers, handleLog)
+	go l.unsubscribeOnDone(done, sub)
+	l.started = true
+	return nil
+}
+
+// Started returns true if the log traversal has started.
+func (l *logTraversal) Started() bool {
+	return l.started
+}
+
+// onNewHead handles a new [types.Header].
+func (l *logTraversal) onNewHead(ctx context.Context, headers chan *types.Header, handleLog func(*types.Log) error) {
+	for {
+		select {
+		case <-l.quit:
+			l.log.Info("stopping log traversal: received quit signal")
+			return
+		case header := <-headers:
+			l.log.Info("received new head", "number", header.Number)
+			l.dispatchNewHead(ctx, header, handleLog, true)
+		}
+	}
+}
+
+// fetchHeaderByNumber gracefully fetches a block header by number with a backoff.
+func (l *logTraversal) fetchHeaderByNumber(ctx context.Context, num *big.Int) (*types.Header, error) {
+	backoffStrategy := l.buildBackoffStrategy()
+	var header *types.Header
+	err := backoff.DoCtx(ctx, 5, backoffStrategy, func() error {
+		ctx, cancel := context.WithTimeout(ctx, time.Second*10)
+		defer cancel()
+		var err error
+		header, err = l.client.HeaderByNumber(ctx, num)
+		return err
+	})
+	return header, err
+}
+
+// spawnCatchup spawns a new goroutine to "catchup" for missed/skipped blocks.
+func (l *logTraversal) spawnCatchup(ctx context.Context, start *big.Int, end *big.Int, handleLog func(*types.Log) error) {
+	for {
+		// Break if we've caught up (start > end).
+		if start.Cmp(end) == 1 {
+			l.log.Info("caught up")
+			return
+		}
+
+		// Fetch the header with a backoff
+		header, err := l.fetchHeaderByNumber(ctx, start)
+		if err != nil {
+			l.log.Error("failed to fetch header", "err", err)
+			return
+		}
+		l.dispatchNewHead(ctx, header, handleLog, false)
```

Comment:
I don't think this is going to work out right because it's including all the checks around `lastBlockNumber` etc but we're backfilling and now have multiple threads accessing and updating `lastBlockNumber`.  Given it already knows the backfill is stepping through each block in order, it just needs the end of `dispatchNewHead` that fetches and processes receipts extracted.

Then it could also just retrieve the full block rather than the header to save on RPC calls.